### DISCUSSION TITLE
No Mapping found error - RequestError: TransportError(400, u'search_phase_execution_exception', u'No mapping found for [alert_time] in order to sort on')

### DISCUSSION THREAD

**Discussion by loGit123** [jertel/elastalert2#1076] [Q&A] [answered]
I am using elastalert2 to get the alerts 

but I am getting this error -No Mapping found- RequestError: TransportError(400, u'search_phase_execution_exception', u'No mapping found for [alert_time] in order to sort on') 

![image](https://user-images.githubusercontent.com/114901513/211165948-1bea298c-5647-42e7-ba4d-aced109c2f61.png)

so tried the following by reading the documentation but it is giving me error

![image](https://user-images.githubusercontent.com/114901513/211165902-a1ac6698-c20b-4a41-b6fc-d402271adfc5.png)
-------
but gives below error -
![image](https://user-images.githubusercontent.com/114901513/211165873-22b1004d-aa0b-415c-8e5b-fd1addf82da9.png)

How to resolve this No mapping found error??

**Comment by jertel** [comment, marked as answer]
Please read #11 and provide the requested information, and properly format your message to make it legible. Once we get this we can try to assist.

**Reply by nsano-rururu** [reply]
https://github.com/jertel/elastalert2/discussions/997

**Reply by loGit123** [reply]
I tried changin the timestamp field to "timestamp", but it still fives error 

![image](https://user-images.githubusercontent.com/114901513/211206558-006d89ed-0f9b-411d-b594-f90d827c1a94.png)

**Reply by jertel** [reply]
What field in your Elastic data index contains the timestamp?

**Reply by loGit123** [reply]
![image](https://user-images.githubusercontent.com/114901513/211208196-4f601213-a2c5-4972-9577-aca976ff504e.png)



alert rule - 


![image](https://user-images.githubusercontent.com/114901513/211207940-be9504d5-5416-4b4f-9bec-fdd5011f754c.png)

**Reply by loGit123** [reply]
I tried removing the timefield stamp from the rule and then it gives error as below -

![image](https://user-images.githubusercontent.com/114901513/211208348-47f11aa5-e728-4467-8901-b839e8a3f832.png)

**Reply by jertel** [reply]
The above image indicates a problem communicating with some external alert target host. From the tiny pieces of information you are sharing it seems you are dealing with multiple problems, possibly all unrelated. Pick one problem at a time, provide all relevant logs, elasticsearch versions, elastalert version, etc, as requested in #11. Let's start there.

**Reply by loGit123** [reply]
I am only facing the problem with this "No Mapping found - @timestamp" , 

![image](https://user-images.githubusercontent.com/114901513/211208905-e4ef97cd-8b54-43a6-9e9c-f531831d42f5.png)

I tried changing the field to "timestamp" but faced the same problem.

![image](https://user-images.githubusercontent.com/114901513/211208939-1e1b6f72-83c7-4f53-9009-33a704ea4bd7.png)


ElasticSearch version

```
{
  "name" : "fed-elastic-0",
  "cluster_name" : "fed-elastic",
  "cluster_uuid" : "VOQZLBzlSxeZFaC__s2QDw",
  "version" : {
    "number" : "7.10.2",
    "build_flavor" : "oss",
    "build_type" : "docker",
    "build_hash" : "747e1cc71def077253878a59143c1f785afa92b9",
    "build_date" : "2021-01-13T00:42:12.435326Z",
    "build_snapshot" : false,
    "lucene_version" : "8.7.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}

```
Elastalert - I am using bitsensor/elastalert latest version 
![image](https://user-images.githubusercontent.com/114901513/211209027-349b2318-7ec8-4998-9fd9-a01032463847.png)

alert yaml file 
``` name: smf-lovepreet-kaur-consul-read-status-alertrule-1
type: any
query_key:
   - event.severity
   - event.podName
   - event.networkFunctionName
index: fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28*
timestamp_field: "@timestamp"
realert:
  minutes: 1
filter:
- query:
    query_string:
      query: "event.eventName: \"Consul Read Status\""


```

**Reply by jertel** [reply]
The error message above states that it cannot find a field named `@timestamp` or `timestamp` in the `fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-20*` index. You need to review that index and it's field mappings, either via cURL or via Kibana, and figure out what field contains the time. In your earlier image you posted a screenshot showing a `@timestamp` field in Kibana, but that index has a different name, and it's possible that the index with the problem was created without the `@timestamp` field. This is really an Elasticsearch issue, not an ElastAlert 2 issue. Further, you've confirmed you're not using ElastAlert 2, which is what this project's discussion forums are intended. You should consider switching to ElastAlert 2 if you want communitysupport going forward, or reach out to the author of the `bitsensor/elastalert` version if you intend to continue using that project instead.

**Reply by loGit123** [reply]
The index I added in the image and in the error is the same as it matches the index with "*" (all that exists) .
```
fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28*
```
As shown in the image, I have @timestamp field for the index but still it not able to map it
![image](https://user-images.githubusercontent.com/114901513/211215152-f54a9b85-2c79-430f-8057-aafc794f424d.png)


I am using elastalert2 by downloading the code of elastalert2 inside the dockerfile of bitsensor/elastalert. 
Is there any change in the elastalert2 done that needs some configuration changes for ELasticSearch?

**Reply by jertel** [reply]
No required changes that I'm aware of. What does the Kibana Index Mapping screen show for that `@timestamp` field?

**Reply by nsano-rururu** [reply]
@loGit123 

bitsensor/elastalert is no longer maintained. Use one of the following docker images that forks bitsensor/elastalert, modifies/adds features, and installs elastalert2.

**karql/elastalert2-server**
https://hub.docker.com/r/karql/elastalert2-server/tags
**praecoapp/elastalert-server**
https://hub.docker.com/r/praecoapp/elastalert-server/tags

**Reply by loGit123** [reply]
This is showing 

I checked mapping by 
```
GET fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09/_mapping
```
![image](https://user-images.githubusercontent.com/114901513/211241711-241573c1-6190-4d99-b261-b72830877dd6.png)

**Reply by nsano-rururu** [reply]
https://github.com/Yelp/elastalert/issues/1252#issuecomment-318714027
>The mapping occasionally fails to apply cleanly, I'm not entirely sure why. The fix for this is to delete the elastalert_status index and try running elastalert-create-index again.

**Reply by loGit123** [reply]
I tried recreating the index but it is giving me error 

```
elastalert-create-index
```

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/requests-2.28.1-py3.10.egg/requests/adapters.py", line 489, in send
    resp = conn.urlopen(
  File "/usr/lib/python3.10/site-packages/urllib3-1.26.13-py3.10.egg/urllib3/connectionpool.py", line 787, in urlopen
    retries = retries.increment(
  File "/usr/lib/python3.10/site-packages/urllib3-1.26.13-py3.10.egg/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='10.164.201.95', port=9200): Max retries exceeded with url: /https://10.164.201.95:9200/ (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:997)')))
```

**Reply by loGit123** [reply]
Is there no solution which will fix this error?

**Reply by nsano-rururu** [reply]
It may not be the same as it can be renamed in elastalert.yaml, but by default the following indexes are subject to deletion.
If you are using kibana, you can delete it from the kibana screen.
If you modify the dockerfile of `bitsensor/elastalert` with docker and run it with docker, if you restart the docker elastalert2 container after deleting the index, the elastalert2 index will be recreated at that time. Because `elastalert-create-index` works

`elastalert_status_status`
`elastalert_status`
`elastalert_status_past`
`elastalert_status_silence`
`elastalert_status_error`

example

```sh
curl -XDELETE localhost:9200/elastalert_status_status?pretty=true
curl -XDELETE localhost:9200/elastalert_status?pretty=true
curl -XDELETE localhost:9200/elastalert_status_past?pretty=true
curl -XDELETE localhost:9200/elastalert_status_silence?pretty=true
curl -XDELETE localhost:9200/elastalert_status_error?pretty=true
```

**Reply by loGit123** [reply]
> It may not be the same as it can be renamed in elastalert.yaml, but by default the following indexes are subject to deletion. If you are using kibana, you can delete it from the kibana screen. If you modify the dockerfile of `bitsensor/elastalert` with docker and run it with docker, if you restart the docker elastalert2 container after deleting the index, the elastalert2 index will be recreated at that time. Because `elastalert-create-index` works
> 
> `elastalert_status_status` `elastalert_status` `elastalert_status_past` `elastalert_status_silence` `elastalert_status_error`
> 
> example
> 
> ```shell
> curl -XDELETE localhost:9200/elastalert_status_status?pretty=true
> curl -XDELETE localhost:9200/elastalert_status?pretty=true
> curl -XDELETE localhost:9200/elastalert_status_past?pretty=true
> curl -XDELETE localhost:9200/elastalert_status_silence?pretty=true
> curl -XDELETE localhost:9200/elastalert_status_error?pretty=true
> ```

Ok. I'll try this .

**Reply by loGit123** [reply]
> [Yelp/elastalert#1252 (comment)](https://github.com/Yelp/elastalert/issues/1252#issuecomment-318714027)
> 
> > The mapping occasionally fails to apply cleanly, I'm not entirely sure why. The fix for this is to delete the elastalert_status index and try running elastalert-create-index again.

No other solution is there ? because deleting/recreating would not be ideal solution I guess.

**Reply by jertel** [reply]
This problem is not concerned with the elastalert_ internal indices. The problem is with the user's data indices. I suspect that one of the daily indices that match the `fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28*` wildcard pattern does not contain a `@timestamp` field. When ElastAlert 2 submits a request to Elasticsearch using a wildcard pattern, ALL matching indices will be searched and sorted. Even if only one of 1000 matching indices fails to have that `@timestamp` field mapped it will cause the request to fail with the error that you are seeing. If you go into Kibana and view the Index pattern, it will usually tell you there's a field conflict.

**Reply by jertel** [reply]
As a quick test, try hardcoding that specific daily index `fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09` into the rule yaml file:

```
index: fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09
```

**Reply by loGit123** [reply]
> As a quick test, try hardcoding that specific daily index `fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09` into the rule yaml file:
> 
> ```
> index: fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09
> ```

I tried by hardcoding the index in rule yaml file
```
spec:
  alertrules: |
    name: smf-lovepreet-kaur-10-164-33-126-01-08-15-28-smfcc-consul-read-status-alertrule-1
    type: any
    query_key:
       - event.severity
       - event.podName
       - event.networkFunctionName
    index: fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09
    timestamp_field: "@timestamp"
    realert:
      minutes: 1
    filter:
    - query:
        query_string:
          query: "event.eventName: \"Consul Read Status\""
    alert:
    - post:

```
then I am not seeing the error of "No Mapping Found" but then Internal Server occurs which is strange as I have configured other alerts and they are able to hit 

```
{"name":"elastalert-server","hostname":"elastalert-645c966fdc-w2n7m","pid":18,"level":50,"msg":"ProcessController:  ERROR:elastalert:Error while running alert http_post: Error posting HTTP Post alert: 500 Server Error: Internal Server Error for url:
```

----------------------------------------------------------------------------------------------------------------
Logs of other alerts that are getting posted successfully which never threw "No mapping error" 

alert rule yaml 
```
spec:
  alertrules: |
    name: smf-lovepreet-kaur-10-164-33-126-01-08-15-28-smfcc-critical-service-not-available-alertrule-1
    type: any
    query_key:
       - event.severity
       - event.criticalServiceName
       - event.networkFunctionName
    index: fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09
    timestamp_field: "@timestamp"
    realert:
      minutes: 1
    filter:
    - query:
        query_string:
          query: "event.eventName: \"Critical Service not Available\""
    alert:
    - post:

```
```
{"name":"elastalert-server","hostname":"elastalert-645c966fdc-w2n7m","pid":18,"level":30,"msg":"Routes:  Successfully handled POST request for '/rules/:id'.","time":"2023-01-09T21:30:20.557Z","v":0}
```

**Reply by loGit123** [reply]
One difference between 

```
Consul read status alert  and Critical Service Unavailable is that
Consul read status severity is "Info" 
and Critical Service Unavailable is "Critical" 
```

and "Info" alert is throwing "no Mapping found" error. What could be the reason?

**Reply by jertel** [reply]
I can't help you without seeing the entire error message and relevant logs. Please include the full log message and surrounding log messages. Also, check Kibana's index mapping page as I suggested earlier, because I suspect you will see an index field conflict error which will tell you that you have a bad index in your cluster.

**Reply by loGit123** [reply]
This is the full error message for when I hardcoded the index - ``` fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09``` 

![image](https://user-images.githubusercontent.com/114901513/211558866-beb147c1-da92-401a-842f-33869a391205.png)

```
Error while running alert http_post: Error posting HTTP Post alert: 500 Server Error: Internal Server Error for url: http://10.164.201.64/api/webhooks/elastalert?api-key=admin
```
Traceback -
```
Traceback (most recent call last):,   File "/opt/elastalert/elastalert/alerters/httppost.py", line 54, in alert,     response.raise_for_status(),   File "/usr/lib/python3.10/site-packages/requests-2.28.1-py3.10.egg/requests/models.py", line 1021, in raise_for_status,     raise HTTPError(http_error_msg, response=self), requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://10.164.201.64/api/webhooks/elastalert?api-key=admin, , During handling of the above exception, another exception occurred:, , Traceback (most recent call last):,   File "/opt/elastalert/elastalert/elastalert.py", line 1375, in send_alert,     alert.alert(matches),   File "/opt/elastalert/elastalert/alerters/httppost.py", line 56, in alert,     raise EAException("Error posting HTTP Post alert: %s" % e), elastalert.util.EAException: Error posting HTTP Post alert: 500 Server Error: Internal Server Error for url: http://10.164.201.64/api/webhooks/elastalert?api-key=admin
```

**Reply by loGit123** [reply]
> I can't help you without seeing the entire error message and relevant logs. Please include the full log message and surrounding log messages. Also, check Kibana's index mapping page as I suggested earlier, because I suspect you will see an index field conflict error which will tell you that you have a bad index in your cluster.

In the kibana's mapping I am not seeing any conflict error.

**Reply by jertel** [reply]
I think we're going in circles. You have pasted an error message showing that ElastAlert 2 is attempting to send an alert, via httppost, to an external server's API (`/api/webhooks/elastalert`). I don't know what that server is, perhaps it's the _bitsensor/elastalert-server_, but it's not part of this project and I can't help you troubleshoot it. That's one problem. 

The other problem, with the index's @timestamp field not existing, is because you have a mixture of indices in your Elasticsearch cluster, where some have the `@timestamp` field, but at least one does not. This is an Elasticsearch problem and you may need to consult with the Elastic discussion forums for assistance if you do not know how to fix this on your own. Most likely you will need to track down which daily index is bad, and remove it or repair it.

Neither of these problems are caused by ElastAlert 2.

**Reply by loGit123** [reply]
Okay, 
These are the logs which I took for "elastalert_status_error" index which yes it is sending an alert using bitsensor/elastalert server only but I am not sure where the problem in occuring because ``` @timestamp``` field is existing in that particular index, and there is only one index but daily it gets created that is why used *(wildcard pattern). 


Thank you for your help!

**Reply by jertel** [reply]
It's not one index. It's one _Index Pattern_, representing many daily indices. Your hard coded test you ran an hour ago proves that the index `fluentd-event-smf-lovepreet-kaur-10-164-33-126-01-08-15-28-2023.01.09` is valid, and correctly contains the `@timestamp` field. It's only when you attempt to use the wildcard that you get the mapping error. 

Now, when you hard coded the good index, ElastAlert 2 was able to successfully run the rule query and collect the results. It did not encounter the mapping error. It then attempted to POST an alert to your bitsensor server, which then failed internally. This is not related to the index mapping error.

This is the best I can do to help you track this down, since I don't have access to your servers.

**Reply by loGit123** [reply]
ok, Thanks!
